### PR TITLE
tp-qemu: add virtio port hotplug/unplug tests

### PR DESCRIPTION
**Pull request by xutian**
Signed-off-by: Xu Tian <xutian@redhat.com>

### CODE REVIEW

**Comment by xutian** [comment]
ID: 1157455 1157384

**Comment by xutian** [comment]
ID: 1157458 1157436

**Comment by xutian** [comment]
[root@dhcp-10-29]# avocado run --vt-type=qemu  --vt-guest-os=RHEL.7.2 --vt-extra-params=nettype=bridge  io-github-autotest-qemu.virtio_port_hotplug
JOB ID     : 061b085cdfeebfcc5e20bd316584a3d47c7b6585
JOB LOG    : /root/avocado/job-results/job-2016-07-25T15.33-061b085/job.log
TESTS      : 2
 (1/2) type_specific.io-github-autotest-qemu.virtio_port_hotplug.unplug_chardev: PASS (174.23 s)
 (2/2) type_specific.io-github-autotest-qemu.virtio_port_hotplug.unplug_serial: PASS (206.16 s)
RESULTS    : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0
JOB HTML   : /root/avocado/job-results/job-2016-07-25T15.33-061b085/html/results.html
TESTS TIME : 380.39 s
No handlers could be found for logger "avocado.test.stdout"

**Comment by suqinhuang** [comment]
https://github.com/avocado-framework/avocado-vt/pull/368

**Comment by suqinhuang** [comment]
Hi @xutian could you share the id

**Comment by suqinhuang** [comment]
always failed to login after vm.reboot()

  File "/usr/lib/python2.7/site-packages/autotest/client/shared/error.py", line 140, in new_fn
    return fn(*args, **kwargs)
  File "/usr/share/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/virtio_port_hotplug.py", line 84, in run
    vm.reboot()
  File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-38.0-py2.7.egg/virttest/error_context.py", line 135, in new_fn
    return fn(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-38.0-py2.7.egg/virttest/qemu_vm.py", line 3920, in reboot
    return self.wait_for_login(nic_index=nic_index, timeout=(timeout / 2))
  File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-38.0-py2.7.egg/virttest/virt_vm.py", line 1040, in wait_for_login
    s_session = self.serial_login(username=username, password=password)
  File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-38.0-py2.7.egg/virttest/error_context.py", line 135, in new_fn
    return fn(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-38.0-py2.7.egg/virttest/virt_vm.py", line 1199, in serial_login
    remote.handle_prompts(console, username, password, prompt, timeout)
  File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-38.0-py2.7.egg/virttest/remote.py", line 193, in handle_prompts

**Comment by xutian** [comment]
which case, can you show me full debug log.

File: qemu/tests/virtio_port_hotplug.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+                port_chardev.unplug(vm.monitor)
+                time.sleep(0.5)
+                port_chardev.hotplug(vm.monitor)
+            virtio_port.hotplug(vm.monitor)
+            if module:
+                error.context("modprobe the module %s" % module, logging.info)
+                session.cmd("modprobe %s" % module)
+        vm.reboot()
+        session = vm.wait_for_login(timeout=timeout)
+        session.close()
```
Comment (CongLi) [inline qemu/tests/virtio_port_hotplug.py]: The 3 lines above should be out of the loop, right?

Comment (xutian) [inline qemu/tests/virtio_port_hotplug.py]: No, it should be in loop.

File: qemu/tests/virtio_port_hotplug.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+            error.context("modporbe the module %s" % module, logging.info)
+            session.cmd("modprobe %s" % module)
+        for port in params.objects("virtio_ports"):
+            port_params = params.object_params(port)
+            port_name = get_virtio_port_name_by_params(port_params, port)
+            virtio_port = get_virtio_port_by_name(vm, port_name)
+            chardev_qid = virtio_port.get_param("chardev")
+            port_chardev = vm.devices.get_by_qid(chardev_qid)[0]
+            if module:
+                error.context("modporbe the module %s" % module, logging.info)
```
Comment (CongLi) [inline qemu/tests/virtio_port_hotplug.py]: The context should be remove module here?

File: qemu/tests/virtio_port_hotplug.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        session = vm.wait_for_login(timeout=timeout)
+        module = params.get("modprobe_module")
+        if module:
+            error.context("modporbe the module %s" % module, logging.info)
+            session.cmd("modprobe %s" % module)
+        for port in params.objects("virtio_ports"):
+            port_params = params.object_params(port)
+            port_name = get_virtio_port_name_by_params(port_params, port)
+            virtio_port = get_virtio_port_by_name(vm, port_name)
+            chardev_qid = virtio_port.get_param("chardev")
```
Comment (CongLi) [inline qemu/tests/virtio_port_hotplug.py]: 10:44:37 ERROR|   File "/usr/share/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/virtio_port_hotplug.py", line 64, in run
10:44:37 ERROR|     chardev_qid = virtio_port.get_param("chardev")
10:44:37 ERROR| AttributeError: 'NoneType' object has no attribute 'get_param'

File: qemu/tests/virtio_port_hotplug.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        repeat += 1
+        session = vm.wait_for_login(timeout=timeout)
+        module = params.get("modprobe_module")
+        if module:
+            error.context("modporbe the module %s" % module, logging.info)
+            session.cmd("modprobe %s" % module)
+        for port in params.objects("virtio_ports"):
+            port_params = params.object_params(port)
+            port_name = get_virtio_port_name_by_params(port_params, port)
+            virtio_port = get_virtio_port_by_name(vm, port_name)
```
Comment (CongLi) [inline qemu/tests/virtio_port_hotplug.py]: 10:49:47 ERROR|   File "/usr/share/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/virtio_port_hotplug.py", line 64, in run
10:49:47 ERROR|     chardev_qid = virtio_port.get_param("chardev")
10:49:47 ERROR| AttributeError: 'NoneType' object has no attribute 'get_param'

**Comment by xutian** [comment]
@CongLi  updated and not found the issue that virtio_port is None.

PASS 1-smp_2.4096m.repeat1.run_test.Host_RHEL.m6.u8.qcow2.virtio_blk.up.virtio_net.RHEL.7.2.x86_64.io-github-autotest-qemu.virtio_port_hotplug.unplug_chardev
PASS 
2-smp_2.4096m.repeat1.run_test.Host_RHEL.m6.u8.qcow2.virtio_blk.up.virtio_net.RHEL.7.2.x86_64.io-github-autotest-qemu.virtio_port_hotplug

thanks,
Xu

**Comment by xutian** [comment]
@suqinhuang  and @CongLi   would you like to review this PR, thanks.

**Comment by CongLi** [comment]
Ack.

**Comment by XueqiangWei** [comment]
LGTM,
Acked-by XueqiangWei <xuwei@redhat.com>

**Comment by ngu-niny** [comment]
@xutian The test failed to get virtio port:

23:09:24 ERROR| Traceback (most recent call last):
23:09:24 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-39.0-py2.7.egg/avocado_vt/test.py", line 196, in runTest
23:09:24 ERROR|     self._runTest()
23:09:24 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-39.0-py2.7.egg/avocado_vt/test.py", line 337, in _runTest
23:09:24 ERROR|     run_func(self, params, env)
23:09:24 ERROR|   File "/usr/lib/python2.7/site-packages/autotest/client/shared/error.py", line 140, in new_fn
23:09:24 ERROR|     return fn(*args, **kwargs)
23:09:24 ERROR|   File "/usr/share/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/virtio_port_hotplug.py", line 71, in run
23:09:24 ERROR|     port_name)
23:09:24 ERROR| TestFail: Virtio Port named 'vc1' not found

**Comment by xutian** [comment]
have you really applied patch avocado-framework/avocado-vt#368 .

File: qemu/tests/virtio_port_hotplug.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+    timeout = int(params.get("login_timeout", 360))
+    for repeat in xrange(int(params.get("repeat_times", 1))):
+        repeat += 1
+        session = vm.wait_for_login(timeout=timeout)
+        module = params.get("modprobe_module")
+        if module:
+            error.context("Load module %s" % module, logging.info)
+            session.cmd("modprobe %s" % module)
```
Comment (ngu-niny) [inline qemu/tests/virtio_port_hotplug.py]: To reduce redundancy, the loading module operation could be moved out of the for loop.

File: qemu/tests/virtio_port_hotplug.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    """
+    Test hot unplug virtio serial devices.
+
+    1) Start guest with virtio serial device(s).
+    2) reload module in guest os.
+    3) Hot-unplug virtio serial port one by one.
+    4) unload module in guest
+    5) Hotplug it one by one.
+    6) reload module in guest.
+    7) Reboot VM to check guest kernel not panic.
```
Comment (ngu-niny) [inline qemu/tests/virtio_port_hotplug.py]: Seems the steps description is not correct, following is better:
1) Start guest with virtio serial device(s).
2) Load module in guest os.
3) For each of the virtio serial ports, do following steps one by one:
    3.1) Unload module in guest
    3.2) Hot-unplug the virtio serial port
    3.3) Hotplug the devices
    3.4) Reload module in the guest
4) Reboot VM to make sure the guest kernel not panic.

**Comment by ngu-niny** [comment]
@xutian After apply patch avocado-framework/avocado-vt#368, the cases could pass.

However, found 2 small issues, I have added comment.

File: qemu/tests/virtio_port_hotplug.py (hunk 5 above)
Comment (xutian) [inline qemu/tests/virtio_port_hotplug.py]: no this step is used to check is virtio-serial module is strange enough, so no need to move out it

**Comment by xutian** [comment]
@ngu-niny  thanks your comments, step info updated, please review. 

Thanks,
Xu

**Comment by xutian** [comment]
@ngu-niny  thanks your comments, step info updated, please review. 

Thanks,
Xu

**Comment by ngu-niny** [comment]
Acked-by: Nini Gu <ngu@redhat.com>

**Comment by xutian** [comment]
thanks @CongLi  and @ngu-niny 's test results

**Pull request merged**